cli/flox-rust-sdk/src/models/env_registry.rs

```diff
@@ -0,0 +1,485 @@
+use std::fs::{File, OpenOptions};
+use std::io::{BufReader, BufWriter};
+use std::path::{Path, PathBuf};
+use std::time::SystemTime;
+
+use fslock::LockFile;
+use serde::{Deserialize, Serialize};
+use tempfile::tempdir_in;
+use tracing::debug;
+
+use super::environment::{path_hash, CanonicalPath, CanonicalizeError, EnvironmentPointer};
+use crate::data::Version;
+use crate::flox::Flox;
+use crate::utils::traceable_path;
+
+pub const ENV_REGISTRY_FILENAME: &str = "env-registry.json";
+
+/// Errors encountered while interacting with the environment registry.
+#[derive(Debug, thiserror::Error)]
+pub enum EnvRegistryError {
+    #[error("couldn't acquire environment registry file lock")]
+    AcquireLock(#[source] fslock::Error),
+    #[error("couldn't open environment registry file")]
+    OpenRegistry(#[source] std::io::Error),
+    #[error("couldn't parse environment registry")]
+    ParseRegistry(#[source] serde_json::Error),
+    #[error("couldn't read environment pointer")]
+    ReadPointer(#[source] std::io::Error),
+    #[error("couldn't parse environment pointer")]
+    ParsePointer(#[source] serde_json::Error),
+    #[error("no '.flox' directory at provided path")]
+    DotFloxDoesntExist(#[source] CanonicalizeError),
+    #[error("failed to open temporary file for registry")]
+    OpenTmpRegistry(#[source] std::io::Error),
+    #[error("failed to write temporary environment registry file")]
+    WriteTmpRegistry(#[source] serde_json::Error),
+    #[error("failed to rename temporary registry file")]
+    RenameRegistry(#[source] std::io::Error),
+    #[error("registry file stored in an invalid location: {0}")]
+    InvalidRegistryLocation(PathBuf),
+    #[error("no environments registered with key: {0}")]
+    UnknownKey(String),
+    #[error("did not find environment in registry")]
+    EnvNotRegistered,
+}
+
+/// A local registry of environments on the system.
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+#[cfg_attr(test, derive(proptest_derive::Arbitrary, PartialEq))]
+pub struct EnvRegistry {
+    /// The schema version of the local environment registry file.
+    pub version: Version<1>,
+    /// The list of locations at which environments can be found and the metadata about
+    /// the environments that have existed there.
+    // Note: We use this ugly macro to generate fewer `RegistryEntry`s than `proptest`s default,
+    //       and this makes a _huge_ difference in test execution speed.
+    #[cfg_attr(
+        test,
+        proptest(
+            strategy = "proptest::collection::vec(proptest::arbitrary::any::<RegistryEntry>(), 0..=3)"
+        )
+    )]
+    pub entries: Vec<RegistryEntry>,
```

Comment:
Conceptually it makes sense for this to be a set, but there's actually no `iter_mut` method on either `HashSet` or `BTreeSet`. This makes sense because you could modify a member of the set to be identical to another member of the set, which breaks an invariant.

Reply:
So, we could make this a set and make operations on `entries` remove the entry while it's being mutated then insert it back into the set, _or_ we could just use a `Vec`.